backend/controller/leader/coordinator.go

```diff
@@ -0,0 +1,173 @@
+package leader
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/TBD54566975/ftl/backend/controller/dal"
+	"github.com/TBD54566975/ftl/backend/controller/leases"
+	"github.com/TBD54566975/ftl/internal/log"
+	"github.com/alecthomas/types/optional"
+)
+
+// LeaderFactory is a function that is called whenever a new leader is acquired.
+//
+// The context provided is tied to the lease and will be cancelled when the leader is no longer leading.
+type LeaderFactory[P any] func(ctx context.Context) (P, error)
+
+// FollowerFactory is a function that is called whenever we follow a new leader.
+//
+// If the new leader has the same url as the previous leader, the existing follower will be used.
+type FollowerFactory[P any] func(ctx context.Context, url *url.URL) (P, error)
+
+type leader[P any] struct {
+	value P
+	lease leases.Lease
+}
+
+type follower[P any] struct {
+	value    P
+	deadline time.Time
+	url      *url.URL
+}
+
+// Coordinator assigns a single leader for the rest to follow.
+// It uses a lease to ensure that only one leader is active at a time.
+type Coordinator[P any] struct {
```

Comment:
Document what P is

Reply:
I assume this is a gRPC client? If so use `rpc.Pingable` to constrain this type.

Reply:
Will document what it is.

It's not a gRPC client btw. It's the protocol that leader and follower conform to.
In the ASM use case it doesn't quite make sense for the leader and follower to conform to a gPRC protocol. Atleast not how I've been thinking about it.